File: src/main/java/com/wire/xenon/WireAPI.java
package com.wire.xenon;

import com.wire.xenon.assets.IAsset;
import com.wire.xenon.backend.models.Conversation;
import com.wire.xenon.backend.models.User;
import com.wire.xenon.exceptions.HttpException;
import com.wire.xenon.models.AssetKey;
import com.wire.xenon.models.otr.*;

import javax.validation.constraints.NotNull;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

public interface WireAPI {
    Devices sendMessage(OtrMessage msg, Object... ignoreMissing) throws HttpException;

    Devices sendPartialMessage(OtrMessage msg, UUID userId) throws HttpException;

    Collection<User> getUsers(Collection<UUID> ids);

    User getSelf();

    Conversation getConversation();

    PreKeys getPreKeys(Missing missing);

    ArrayList<Integer> getAvailablePrekeys(@NotNull String client);

    void uploadPreKeys(ArrayList<PreKey> preKeys) throws IOException;

    AssetKey uploadAsset(IAsset asset) throws Exception;

    byte[] downloadAsset(String assetId, String assetToken) throws HttpException;

    boolean deleteConversation(UUID teamId) throws HttpException;

    User addService(UUID serviceId, UUID providerId) throws HttpException;

    User addParticipants(UUID... userIds) throws HttpException;

    Conversation createConversation(String name, UUID teamId, List<UUID> users) throws HttpException;

    Conversation createOne2One(UUID teamId, UUID userId) throws HttpException;

    void leaveConversation(UUID user) throws HttpException;

    User getUser(UUID userId) throws HttpException;

    UUID getUserId(String handle) throws HttpException;

    boolean hasDevice(UUID userId, String clientId);

    UUID getTeam() throws HttpException;

    Collection<UUID> getTeamMembers(UUID teamId);

    void acceptConnection(UUID user) throws Exception;
}

File: src/main/java/com/wire/kalium/WireAPI.kt
package com.wire.kalium

import com.wire.kalium.assets.Asset
import com.wire.kalium.backend.models.Conversation
import com.wire.kalium.backend.models.User
import com.wire.kalium.exceptions.HttpException
import com.wire.kalium.models.AssetKey
import com.wire.kalium.models.otr.*
import java.io.IOException
import java.util.*

interface WireAPI {
    @Throws(HttpException::class)
    open fun sendMessage(msg: OtrMessage, vararg ignoreMissing: Any?): Devices
    @Throws(HttpException::class)
    open fun sendPartialMessage(msg: OtrMessage, userId: UUID): Devices
    open fun getUsers(ids: MutableCollection<UUID>): MutableCollection<User>
    open fun getSelf(): User
    open fun getConversation(): Conversation
    open fun getPreKeys(missing: Missing): PreKeys
    open fun getAvailablePrekeys(client: String): ArrayList<Int>
    @Throws(IOException::class)
    open fun uploadPreKeys(preKeys: ArrayList<PreKey>)
    @Throws(Exception::class)
    open fun uploadAsset(asset: Asset): AssetKey
    @Throws(HttpException::class)
    open fun downloadAsset(assetId: String, assetToken: String?): ByteArray
    @Throws(HttpException::class)
    open fun deleteConversation(teamId: UUID): Boolean
    @Throws(HttpException::class)
    open fun addService(serviceId: UUID, providerId: UUID): User
    @Throws(HttpException::class)
    open fun addParticipants(vararg userIds: UUID): User
    @Throws(HttpException::class)
    open fun createConversation(name: String, teamId: UUID, users: MutableList<UUID>): Conversation
    @Throws(HttpException::class)
    open fun createOne2One(teamId: UUID, userId: UUID): Conversation
    @Throws(HttpException::class)
    open fun leaveConversation(user: UUID)
    @Throws(HttpException::class)
    open fun getUser(userId: UUID): User
    @Throws(HttpException::class)
    open fun getUserId(handle: String): UUID
    open fun hasDevice(userId: UUID, clientId: String): Boolean

    @Throws(HttpException::class)
    open fun getTeam(): UUID?
    open fun getTeamMembers(teamId: UUID): MutableCollection<UUID>
    @Throws(Exception::class)
    open fun acceptConnection(user: UUID)
}
